File: backups/20250728_121658/backend_src_backup/services/embeddingService.ts
import OpenAI from 'openai';
import { db, runQuery, runSingleQuery } from '../database';

// Initialize OpenAI client only if API key is available
let openai: OpenAI | null = null;
if (process.env.OPENAI_API_KEY) {
  openai = new OpenAI({
    apiKey: process.env.OPENAI_API_KEY,
  });
}

export interface EmbeddingResult {
  idea_id: number;
  similarity: number;
  title: string;
  description?: string;
  category?: string;
  status: string;
}

export class EmbeddingService {
  // Generate embedding for a text using OpenAI
  async generateEmbedding(text: string): Promise<number[]> {
    if (!openai) {
      throw new Error('OpenAI API key not configured');
    }

    try {
      const response = await openai.embeddings.create({
        model: 'text-embedding-3-small',
        input: text,
        encoding_format: 'float',
      });

      return response.data[0].embedding;
    } catch (error) {
      console.error('Error generating embedding:', error);
      throw new Error('Failed to generate embedding');
    }
  }

  // Calculate cosine similarity between two vectors
  calculateCosineSimilarity(vecA: number[], vecB: number[]): number {
    if (vecA.length !== vecB.length) {
      throw new Error('Vectors must have the same length');
    }

    let dotProduct = 0;
    let normA = 0;
    let normB = 0;

    for (let i = 0; i < vecA.length; i++) {
      dotProduct += vecA[i] * vecB[i];
      normA += vecA[i] * vecA[i];
      normB += vecB[i] * vecB[i];
    }

    normA = Math.sqrt(normA);
    normB = Math.sqrt(normB);

    if (normA === 0 || normB === 0) {
      return 0;
    }

    return dotProduct / (normA * normB);
  }

  // Store embedding for an idea
  async storeEmbedding(ideaId: number, embedding: number[]): Promise<void> {
    const embeddingJson = JSON.stringify(embedding);
    
    try {
      await runSingleQuery(
        'INSERT OR REPLACE INTO embeddings (idea_id, embedding, updated_at) VALUES (?, ?, CURRENT_TIMESTAMP)',
        [ideaId, embeddingJson]
      );
    } catch (error) {
      console.error('Error storing embedding:', error);
      throw new Error('Failed to store embedding');
    }
  }

  // Get embedding for an idea
  async getEmbedding(ideaId: number): Promise<number[] | null> {
    try {
      const result = await runQuery(
        'SELECT embedding FROM embeddings WHERE idea_id = ?',
        [ideaId]
      );

      if (result.length === 0) {
        return null;
      }

      return JSON.parse(result[0].embedding);
    } catch (error) {
      console.error('Error getting embedding:', error);
      return null;
    }
  }

  // Generate and store embedding for an idea
  async generateAndStoreEmbedding(ideaId: number, title: string, description?: string, content?: string): Promise<void> {
    if (!openai) {
      console.log('⚠️  Skipping embedding generation - OpenAI API key not configured');
      return;
    }

    // Combine title, description, and content for embedding
    const textToEmbed = [title, description, content].filter(Boolean).join(' ');
    
    if (!textToEmbed.trim()) {
      throw new Error('No text content to embed');
    }

    const embedding = await this.generateEmbedding(textToEmbed);
    await this.storeEmbedding(ideaId, embedding);
  }

  // Find related ideas using semantic similarity
  async findRelatedIdeas(ideaId: number, limit: number = 5): Promise<EmbeddingResult[]> {
    if (!openai) {
      console.log('⚠️  Related ideas feature disabled - OpenAI API key not configured');
      return [];
    }

    try {
      // Get the target idea's embedding
      const targetEmbedding = await this.getEmbedding(ideaId);
      if (!targetEmbedding) {
        return [];
      }

      // Get all other ideas with their embeddings
      const ideasWithEmbeddings = await runQuery(`
        SELECT 
          i.id,
          i.title,
          i.description,
          i.category,
          i.status,
          e.embedding
        FROM ideas i
        INNER JOIN embeddings e ON i.id = e.idea_id
        WHERE i.id != ?
      `, [ideaId]);

      // Calculate similarities
      const similarities: EmbeddingResult[] = [];

      for (const idea of ideasWithEmbeddings) {
        const ideaEmbedding = JSON.parse(idea.embedding);
        const similarity = this.calculateCosineSimilarity(targetEmbedding, ideaEmbedding);

        similarities.push({
          idea_id: idea.id,
          similarity,
          title: idea.title,
          description: idea.description,
          category: idea.category,
          status: idea.status,
        });
      }

      // Sort by similarity (highest first) and return top results
      return similarities
        .sort((a, b) => b.similarity - a.similarity)
        .slice(0, limit);
    } catch (error) {
      console.error('Error finding related ideas:', error);
      return [];
    }
  }

  // Update embeddings for all ideas (useful for initial setup)
  async updateAllEmbeddings(): Promise<void> {
    if (!openai) {
      console.log('⚠️  Skipping embedding updates - OpenAI API key not configured');
      return;
    }

    try {
      const ideas = await runQuery('SELECT id, title, description, content FROM ideas');
      
      for (const idea of ideas) {
        try {
          await this.generateAndStoreEmbedding(
            idea.id,
            idea.title,
            idea.description,
            idea.content
          );
          console.log(`✅ Updated embedding for idea ${idea.id}: ${idea.title}`);
        } catch (error) {
          console.error(`❌ Failed to update embedding for idea ${idea.id}:`, error);
        }
      }
    } catch (error) {
      console.error('Error updating all embeddings:', error);
      throw error;
    }
  }
}

export const embeddingService = new EmbeddingService(); 

File: api/app/services/embedding_service.py
import json
import numpy as np
import requests
from typing import List, Optional
from sqlalchemy.orm import Session
from app.models.idea import Idea, Embedding
import logging

logger = logging.getLogger(__name__)

class OllamaEmbeddingService:
    def __init__(self, model_name: str = "all-minilm", base_url: str = "http://localhost:11434"):
        self.model_name = model_name
        self.base_url = base_url
        self.embedding_dimension = 384  # Dimension for all-MiniLM-L6-v2
        
    def _call_ollama_api(self, text: str) -> Optional[List[float]]:
        """Call Ollama API to generate embeddings"""
        try:
            url = f"{self.base_url}/api/embeddings"
            payload = {
                "model": self.model_name,
                "prompt": text
            }
            
            response = requests.post(url, json=payload, timeout=30)
            response.raise_for_status()
            
            data = response.json()
            if "embedding" in data:
                return data["embedding"]
            else:
                logger.error(f"Unexpected response format from Ollama: {data}")
                return None
                
        except requests.exceptions.RequestException as e:
            logger.error(f"Failed to call Ollama API: {e}")
            return None
        except Exception as e:
            logger.error(f"Unexpected error in embedding generation: {e}")
            return None
    
    def generate_embedding(self, text: str) -> Optional[List[float]]:
        """Generate embedding for given text"""
        if not text or not text.strip():
            return None
            
        # Clean and prepare text
        cleaned_text = text.strip()
        
        # Generate embedding
        embedding = self._call_ollama_api(cleaned_text)
        
        if embedding:
            # Normalize the embedding
            embedding_array = np.array(embedding)
            normalized_embedding = embedding_array / np.linalg.norm(embedding_array)
            return normalized_embedding.tolist()
        
        return None
    
    def calculate_similarity(self, embedding1: List[float], embedding2: List[float]) -> float:
        """Calculate cosine similarity between two embeddings"""
        try:
            vec1 = np.array(embedding1)
            vec2 = np.array(embedding2)
            
            # Calculate cosine similarity
            dot_product = np.dot(vec1, vec2)
            norm1 = np.linalg.norm(vec1)
            norm2 = np.linalg.norm(vec2)
            
            if norm1 == 0 or norm2 == 0:
                return 0.0
                
            similarity = dot_product / (norm1 * norm2)
            return float(similarity)
            
        except Exception as e:
            logger.error(f"Error calculating similarity: {e}")
            return 0.0
    
    def get_idea_text_for_embedding(self, idea: Idea) -> str:
        """Extract and combine text from idea for embedding generation"""
        text_parts = []
        
        if idea.title:
            text_parts.append(idea.title)
        
        if idea.description:
            text_parts.append(idea.description)
            
        if idea.content:
            text_parts.append(idea.content)
            
        # Add tags as context
        if idea.tags:
            tag_names = [tag.name for tag in idea.tags]
            text_parts.append(f"Tags: {', '.join(tag_names)}")
            
        # Add category as context
        if idea.category:
            text_parts.append(f"Category: {idea.category}")
            
        return " | ".join(text_parts)
    
    def update_idea_embedding(self, db: Session, idea: Idea) -> bool:
        """Generate and store embedding for an idea"""
        try:
            # Get text for embedding
            idea_text = self.get_idea_text_for_embedding(idea)
            
            if not idea_text:
                logger.warning(f"No text content found for idea {idea.id}")
                return False
            
            # Generate embedding
            embedding_vector = self.generate_embedding(idea_text)
            
            if not embedding_vector:
                logger.error(f"Failed to generate embedding for idea {idea.id}")
                return False
            
            # Store in database
            db_embedding = db.query(Embedding).filter(Embedding.idea_id == idea.id).first()
            
            if db_embedding:
                # Update existing embedding
                db_embedding.embedding = json.dumps(embedding_vector)
            else:
                # Create new embedding
                db_embedding = Embedding(
                    idea_id=idea.id,
                    embedding=json.dumps(embedding_vector)
                )
                db.add(db_embedding)
            
            db.commit()
            logger.info(f"Successfully updated embedding for idea {idea.id}")
            return True
            
        except Exception as e:
            logger.error(f"Error updating embedding for idea {idea.id}: {e}")
            db.rollback()
            return False
    
    def get_similar_ideas(self, db: Session, idea_id: int, limit: int = 5, min_similarity: float = 0.3) -> List[dict]:
        """Find similar ideas based on embedding similarity"""
        try:
            # Get the target idea and its embedding
            target_idea = db.query(Idea).filter(Idea.id == idea_id).first()
            if not target_idea:
                return []
            
            target_embedding = db.query(Embedding).filter(Embedding.idea_id == idea_id).first()
            if not target_embedding:
                # Generate embedding if it doesn't exist
                if not self.update_idea_embedding(db, target_idea):
                    return []
                target_embedding = db.query(Embedding).filter(Embedding.idea_id == idea_id).first()
            
            target_vector = json.loads(target_embedding.embedding)
            
            # Get all other ideas with embeddings
            other_embeddings = db.query(Embedding).filter(Embedding.idea_id != idea_id).all()
            
            similarities = []
            for emb in other_embeddings:
                try:
                    other_vector = json.loads(emb.embedding)
                    similarity = self.calculate_similarity(target_vector, other_vector)
                    
                    if similarity >= min_similarity:
                        # Get the idea details
                        other_idea = db.query(Idea).filter(Idea.id == emb.idea_id).first()
                        if other_idea:
                            similarities.append({
                                "idea_id": other_idea.id,
                                "similarity": similarity,
                                "title": other_idea.title,
                                "description": other_idea.description,
                                "category": other_idea.category,
                                "status": other_idea.status
                            })
                except Exception as e:
                    logger.error(f"Error calculating similarity for idea {emb.idea_id}: {e}")
                    continue
            
            # Sort by similarity and return top results
            similarities.sort(key=lambda x: x["similarity"], reverse=True)
            return similarities[:limit]
            
        except Exception as e:
            logger.error(f"Error getting similar ideas: {e}")
            return []
    
    def update_all_embeddings(self, db: Session) -> dict:
        """Update embeddings for all ideas"""
        try:
            ideas = db.query(Idea).all()
            success_count = 0
            error_count = 0
            
            for idea in ideas:
                if self.update_idea_embedding(db, idea):
                    success_count += 1
                else:
                    error_count += 1
            
            return {
                "success": True,
                "total_ideas": len(ideas),
                "successful_updates": success_count,
                "failed_updates": error_count
            }
            
        except Exception as e:
            logger.error(f"Error updating all embeddings: {e}")
            return {
                "success": False,
                "error": str(e)
            }

# Global instance
embedding_service = OllamaEmbeddingService() 